api_tests/institutions/views/test_institution_auth.py

```diff
@@ -0,0 +1,79 @@
+from nose.tools import *  # flake8: noqa
+import json
+
+import jwt
+import jwe
+
+from modularodm import Q
+
+from tests.base import ApiTestCase
+from tests.factories import InstitutionFactory, UserFactory
+
+from api.base import settings
+from api.base.settings.defaults import API_BASE
+from website.models import Institution, Node, User
+
+
+class TestInstitutionAuth(ApiTestCase):
+    def setUp(self):
+        super(TestInstitutionAuth, self).setUp()
+        self.institution = InstitutionFactory()
+        self.institution.domains.append('uva.edu')
+        self.institution.metadata_request_fields['fullname'] = 'nameFull'
+        self.institution.save()
+        self.url = '/{0}institutions/auth/'.format(API_BASE)
+
+    def tearDown(self):
+        super(TestInstitutionAuth, self).tearDown()
+        self.institution.remove()
+        User.remove()
+
+    def build_payload(self, username):
+        return jwe.encrypt(jwt.encode({
+            'sub': username,
+            'data': json.dumps({
+                self.institution.metadata_request_fields['fullname']: 'Mr Moco'
+            }),
+        }, settings.JWT_SECRET, algorithm='HS256'), settings.JWE_SECRET)
+
+    def test_creates_user(self):
+        username = 'hmoco@{}'.format(self.institution.domains[0])
+
+        assert_equal(User.find(Q('username', 'eq', username)).count(), 0)
+
+        res = self.app.post(self.url, self.build_payload(username))
+        assert_equal(res.status_code, 204)
+
+        user = User.find_one(Q('username', 'eq', username))
+
+        assert_true(user)
+        assert_in(self.institution, user.affiliated_institutions)
+
+    def test_adds_institution(self):
+        username = 'hmoco@{}'.format(self.institution.domains[0])
+
+        user = User(username=username, fullname='Mr Moco')
+        user.save()
+
+        res = self.app.post(self.url, self.build_payload(username))
+        assert_equal(res.status_code, 204)
+
+        user.reload()
+        assert_in(self.institution, user.affiliated_institutions)
+
+    def test_finds_user(self):
+        username = 'hmoco@{}'.format(self.institution.domains[0])
+
+        user = User(username=username, fullname='Mr Moco')
+        user.affiliated_institutions.append(self.institution)
+        user.save()
+
+        res = self.app.post(self.url, self.build_payload(username))
+        assert_equal(res.status_code, 204)
+
+        user.reload()
+        assert_equal(len(user.affiliated_institutions), 1)
+
+    def test_bad_token(self):
+        res = self.app.post(self.url, 'al;kjasdfljadf', expect_errors=True)
+        assert_equal(res.status_code, 403)
```

Comment:
Good tests, maybe add one that tests the update of a users ` institution_metadata `